sound/soc/soc-dapm.c

```diff
@@ -3068,23 +3069,42 @@ int snd_soc_dapm_new_widgets(struct snd_soc_card *card)
 		case snd_soc_dapm_switch:
 		case snd_soc_dapm_mixer:
 		case snd_soc_dapm_mixer_named_ctl:
-			dapm_new_mixer(w);
+			ret = dapm_new_mixer(w);
```

Comment:
Checking the return is good, but dapm_new_mixer, dapm_new_mux etc should free any resources they allocate on failure. likewise dapm_create_or_share_kcontrol() will also need to be check that it frees any resource it allocates on failure. So please

1. Check return values and propagate the return values up the call stack.
2. Free any resources in the function that allocates them on any failure.

Reply:
1. I checked the code in the dapm_create_or_share_kcontrol(), when it detected the error and before return, it has already finished free operation. in this functin, it has only two allocation operation for long_name and kcontrol.
like the snd_soc_dapm_add_path(), it also do this job already when return with error.
2. we only care about the w->kcontrols itself. when the sub-function return error. it will free the allocated w->kcontrols.
3. But the updated PR also has the problem: when w->num_kcontrols is not equal to 0. but the code goto the default case. it will cause the w->kcontrols will be allocated. but never be processed further.
I think we have to free it when enter the default case.